tests/codex/sales/testsales.nim

```diff
@@ -4,157 +4,192 @@ import std/sugar
 import std/times
 import pkg/asynctest
 import pkg/chronos
+import pkg/datastore
+import pkg/questionable
+import pkg/questionable/results
 import pkg/codex/sales
 import pkg/codex/sales/salesdata
+import pkg/codex/sales/salescontext
+import pkg/codex/sales/reservations
+import pkg/codex/stores/repostore
 import pkg/codex/proving
+import pkg/codex/blocktype as bt
 import ../helpers/mockmarket
 import ../helpers/mockclock
 import ../helpers/eventually
 import ../examples
+import ./helpers
 
 suite "Sales":
 
-  let availability = Availability.init(
-    size=100.u256,
-    duration=60.u256,
-    minPrice=600.u256
-  )
-  var request = StorageRequest(
-    ask: StorageAsk(
-      slots: 4,
-      slotSize: 100.u256,
-      duration: 60.u256,
-      reward: 10.u256,
-    ),
-    content: StorageContent(
-      cid: "some cid"
-    ),
-    expiry: (getTime() + initDuration(hours=1)).toUnix.u256
-  )
   let proof = exampleProof()
 
+  var availability: Availability
+  var request: StorageRequest
   var sales: Sales
   var market: MockMarket
   var clock: MockClock
   var proving: Proving
+  var reservations: Reservations
+  var repo: RepoStore
 
   setup:
+    availability = Availability.init(
+      size=100.u256,
+      duration=60.u256,
+      minPrice=600.u256
+    )
+    request = StorageRequest(
+      ask: StorageAsk(
+        slots: 4,
+        slotSize: 100.u256,
+        duration: 60.u256,
+        reward: 10.u256,
+      ),
+      content: StorageContent(
+        cid: "some cid"
+      ),
+      expiry: (getTime() + initDuration(hours=1)).toUnix.u256
+    )
+
     market = MockMarket.new()
     clock = MockClock.new()
     proving = Proving.new()
-    sales = Sales.new(market, clock, proving)
+    let repoDs = SQLiteDatastore.new(Memory).tryGet()
+    let metaDs = SQLiteDatastore.new(Memory).tryGet()
+    repo = RepoStore.new(repoDs, metaDs)
+    await repo.start()
+    sales = Sales.new(market, clock, proving, repo)
+    reservations = sales.context.reservations
     sales.onStore = proc(request: StorageRequest,
                          slot: UInt256,
-                         availability: ?Availability) {.async.} =
-      discard
+                         onBatch: BatchProc): Future[?!void] {.async.} =
+      return success()
     proving.onProve = proc(slot: Slot): Future[seq[byte]] {.async.} =
       return proof
     await sales.start()
     request.expiry = (clock.now() + 42).u256
 
   teardown:
+    await repo.stop()
     await sales.stop()
 
-  test "has no availability initially":
-    check sales.available.len == 0
+  proc getAvailability: ?!Availability =
+    waitFor reservations.get(availability.id)
```

Comment:
I'd avoid using `waitFor` here. We're already in an async context, so we can use `await`.

Reply:
This was done for a few reasons:
1. In the tests, we're not going to maximise our use of the async event loop
2. The test check is cleaner when `getAvailability` is sync
3. Questionable doesn't seem to support chaining an async proc, eg

This works:
```nim
proc getAvailability: ?!Availability =
  waitFor reservations.get(availability.id)

check getAvailability().?used == success false
```

This doesn't work:
```nim
proc getAvailability: Future[?!Availability] {.async.} =
  return await reservations.get(availability.id)

check (await getAvailability()).?used == success false
```
Errors with:
```shell
/Users/egonat/repos/status-im/nim-codex/vendor/questionable/questionable/chaining.nim(21, 20) Error: undeclared field: 'unsafeGet' for type asyncloop.Future [type declared in /Users/egonat/repos/status-im/nim-codex/vendor/nim-chronos/chronos/asyncfutures2.nim(56, 3)]
```